### ISSUE TITLE
`x.py check library/std` does not work analogously to `x.py build library/std`

### ISSUE THREAD

**Issue by varkor** [rust-lang/rust#78466]
I expect these two commands to do the same thing, but for `check` or `build`, respectively. However, `check` may succeed even if `build` fails type checking. I suspect the syntax for `check` is a subset of that for `build`.

**Comment by jyn514** [comment]
@varkor can you give an example of code that fails for check but succeeds for build?

**Comment by varkor** [comment]
I edited some code in `compiler/rustc_middle/src/infer/canonical.rs`. This had a type-checking error in it, but `x.py check library/std` succeeded. When I ran `x.py check` (without specifying a path), it emitted an error as expected. `x.py build library/std` emitted an error with the path specified.

**Comment by jyn514** [comment]
@varkor I think this is actually related to staging: `x.py build library/std` builds the stage0 compiler artifacts before building stage1 libstd with the newly built compiler. `x.py check` always uses stage 0, so `x.py check library/std` doesn't build the compiler.

**Comment by varkor** [comment]
@jyn514: `x.py check` (without the path) emitted the correct error. Does this mean the `check` stage is affected by the path?

**Comment by jyn514** [comment]
@varkor yes, the same as build. If you run x.py compiler/rustc for example, that won't check rustdoc.

**Comment by varkor** [comment]
@jyn514: are you saying that `x.py check library/std` will literally only check `library/std`, using the definitions from the stage0 artefacts, without taking account of the current compiler changes (i.e. those which would be built for stage1)?

**Comment by jyn514** [comment]
@varkor yes: `x.py check library/std` is roughly analogous to `RUSTFLAGS=--profile=check x.py build --stage 0 library/std`. There's not currently a way to check with stage1, although I'd like to add one: https://github.com/rust-lang/rust/issues/46955#issuecomment-753264720

**Comment by varkor** [comment]
Okay, I see. I don't necessarily think there needs to be an option to change the stage: it should just match whatever `build` does (though this will also take some work to implement).

**Comment by jyn514** [comment]
@varkor well, if it matches `build` then you'd have to build the entire compiler before changing the stage 1 standard library, which would double or triple the time. So I don't think stage 1 should be the default.

**Comment by jyn514** [comment]
(continued on Zulip so there's lower latency: https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/x.2Epy.20check.20stage.200)

**Comment by jyn514** [comment]
We decided the best approach is to warn on `x.py check library/std` if you have `profile = "compiler"` set (and recommend `x.py check` instead).

**Comment by jyn514** [comment]
I changed the title since https://github.com/rust-lang/rust/issues/46955 already tracks --stage 1 support for `x.py check`.

**Comment by jyn514** [comment]
This would be helped by my proposal in https://jyn.dev/2023/01/12/Bootstrapping-Rust-in-2023.html.
